error messages in crypto plugins&tests

I think in general it would be good if the crypto plugins and the crypto test cases give better error messages. Should I open a new issue for this? 

```
 95/267 Test  #93: testmod_gpgme .................................***Failed    0.18 sec
OK closing connection
GPGME        TESTS
==================

/home/markus/Projekte/Elektra/current/src/plugins/gpgme/testmod_gpgme.c:227: error in test_encryption_decryption: kdb get failed
/home/markus/Projekte/Elektra/current/src/plugins/gpgme/testmod_gpgme.c:243: error in test_encryption_decryption: key type not restored to string during decryption
/home/markus/Projekte/Elektra/current/src/plugins/gpgme/testmod_gpgme.c:244: error in test_encryption_decryption: key content not restored during decryption
/home/markus/Projekte/Elektra/current/src/plugins/gpgme/testmod_gpgme.c:254: error in test_encryption_decryption: key content not restored during decryption

gpgme Results: 45 Tests done — 4 errors
```

--- a/doc/news/_preparation_next_release.md
+++ b/doc/news/_preparation_next_release.md
@@ -67,6 +67,7 @@ The following section lists news about the [modules](https://www.libelektra.org/
 - Fixed some typos and links in the documentation and add new iterate example. _(Philipp Gackstatter)_
 - We removed `keyRel` and `keyRel2` since it can be easily replaced by other existing functions. _(Philipp Gackstatter)_
 - We improved the error messages in `gpgme`. _(Peter Nirschl)_
+- We improved the error messages in `fcrypt`. _(Peter Nirschl)_
 
 ### Camel
 

--- a/doc/news/_preparation_next_release.md
+++ b/doc/news/_preparation_next_release.md
@@ -66,8 +66,7 @@ The following section lists news about the [modules](https://www.libelektra.org/
 - We unified the name of the config check function of the plugins to `nameOfPluginCheckConf`. Before this update some plugins used the name `nameOfPluginCheckConfig` instead. _(René Schwaiger)_
 - Fixed some typos and links in the documentation and add new iterate example. _(Philipp Gackstatter)_
 - We removed `keyRel` and `keyRel2` since it can be easily replaced by other existing functions. _(Philipp Gackstatter)_
-- We improved the error messages in `gpgme`. _(Peter Nirschl)_
-- We improved the error messages in `fcrypt`. _(Peter Nirschl)_
+- We improved the error messages in `crypto`, `fcrypt`, and `gpgme` plugins. _(Peter Nirschl)_
 
 ### Camel
 

--- a/doc/news/_preparation_next_release.md
+++ b/doc/news/_preparation_next_release.md
@@ -66,6 +66,7 @@ The following section lists news about the [modules](https://www.libelektra.org/
 - We unified the name of the config check function of the plugins to `nameOfPluginCheckConf`. Before this update some plugins used the name `nameOfPluginCheckConfig` instead. _(René Schwaiger)_
 - Fixed some typos and links in the documentation and add new iterate example. _(Philipp Gackstatter)_
 - We removed `keyRel` and `keyRel2` since it can be easily replaced by other existing functions. _(Philipp Gackstatter)_
+- We improved the error messages in `gpgme`. _(Peter Nirschl)_
 
 ### Camel
 

--- a/src/plugins/crypto/botan_operations.cpp
+++ b/src/plugins/crypto/botan_operations.cpp
@@ -152,7 +152,7 @@ int elektraCryptoBotanInit (Key * errorKey)
 	}
 	catch (std::exception const & e)
 	{
-		ELEKTRA_SET_PLUGIN_MISBEHAVIOR_ERRORF (errorKey, "Botan initialization failed. Reason: %s", e.what ());
+		ELEKTRA_SET_INSTALLATION_ERRORF (errorKey, "Botan initialization failed. Reason: %s", e.what ());
 		return -1; // failure
 	}
 	return 1; // success

--- a/src/plugins/crypto/gcrypt_operations.c
+++ b/src/plugins/crypto/gcrypt_operations.c
@@ -233,8 +233,7 @@ int elektraCryptoGcryHandleCreate (elektraCryptoHandle ** handle, KeySet * confi
 error:
 	memset (keyBuffer, 0, sizeof (keyBuffer));
 	memset (ivBuffer, 0, sizeof (ivBuffer));
-	// TODO: Correct??
-	ELEKTRA_SET_INTERNAL_ERRORF (errorKey, "Failed to create handle. Reason: %s", gcry_strerror (gcry_err));
+	ELEKTRA_SET_INTERNAL_ERRORF (errorKey, "Failed to setup libgcrypt. Reason: %s", gcry_strerror (gcry_err));
 	gcry_cipher_close (**handle);
 	elektraFree (*handle);
 	(*handle) = NULL;
@@ -317,8 +316,7 @@ int elektraCryptoGcryEncrypt (elektraCryptoHandle * handle, Key * k, Key * error
 	gcry_err = gcry_cipher_encrypt (*handle, current, ELEKTRA_CRYPTO_GCRY_BLOCKSIZE, NULL, 0);
 	if (gcry_err != 0)
 	{
-		// TODO: Correct??
-		ELEKTRA_SET_INSTALLATION_ERRORF (errorKey, "Encryption failed. Reason: %s", gcry_strerror (gcry_err));
+		ELEKTRA_SET_INTERNAL_ERRORF (errorKey, "Encryption failed. Reason: %s", gcry_strerror (gcry_err));
 		memset (output, 0, outputLen);
 		elektraFree (output);
 		elektraFree (salt);
@@ -333,8 +331,7 @@ int elektraCryptoGcryEncrypt (elektraCryptoHandle * handle, Key * k, Key * error
 	gcry_err = gcry_cipher_encrypt (*handle, current, dataLen, NULL, 0);
 	if (gcry_err != 0)
 	{
-		// TODO: Correct??
-		ELEKTRA_SET_INSTALLATION_ERRORF (errorKey, "Encryption failed. Reason: %s", gcry_strerror (gcry_err));
+		ELEKTRA_SET_INTERNAL_ERRORF (errorKey, "Encryption failed. Reason: %s", gcry_strerror (gcry_err));
 		memset (output, 0, outputLen);
 		elektraFree (output);
 		elektraFree (salt);

--- a/src/plugins/crypto/helper.c
+++ b/src/plugins/crypto/helper.c
@@ -155,8 +155,8 @@ int ELEKTRA_PLUGIN_FUNCTION (getSaltFromPayload) (Key * errorKey, Key * k, kdb_o
 	// validate payload length
 	if ((size_t) payloadLen < sizeof (size_t) || payloadLen < 0)
 	{
-		// TODO: Correct??
-		ELEKTRA_SET_INTERNAL_ERRORF (errorKey, "Payload is too small to contain a salt (payload length is: %zu)", payloadLen);
+		ELEKTRA_SET_VALIDATION_SEMANTIC_ERRORF (errorKey, "Payload is too small to contain a salt (payload length is: %zu)",
+							payloadLen);
 		if (salt) *salt = NULL;
 		return -1;
 	}
@@ -172,9 +172,8 @@ int ELEKTRA_PLUGIN_FUNCTION (getSaltFromPayload) (Key * errorKey, Key * k, kdb_o
 	// validate restored salt length
 	if (restoredSaltLen < 1 || restoredSaltLen > (payloadLen - headerLen))
 	{
-		// TODO: Correct??
-		ELEKTRA_SET_INTERNAL_ERRORF (errorKey, "Restored salt has invalid length of %u (payload length is: %zu)", restoredSaltLen,
-					     payloadLen);
+		ELEKTRA_SET_VALIDATION_SEMANTIC_ERRORF (errorKey, "Restored salt has invalid length of %u (payload length is: %zu)",
+							restoredSaltLen, payloadLen);
 		if (salt) *salt = NULL;
 		return -1;
 	}
@@ -226,7 +225,6 @@ kdb_unsigned_long_t ELEKTRA_PLUGIN_FUNCTION (getIterationCount) (Key * errorKey,
 		}
 		else
 		{
-			// TODO: Correct?
 			ELEKTRA_ADD_INSTALLATION_WARNING (errorKey, "Iteration count provided at " ELEKTRA_CRYPTO_PARAM_ITERATION_COUNT
 								    " is invalid. Using default value instead.");
 		}
@@ -274,7 +272,7 @@ int ELEKTRA_PLUGIN_FUNCTION (gpgEncryptMasterPassword) (KeySet * conf, Key * err
 	if (recipientCount == 0)
 	{
 		char * errorDescription = ELEKTRA_PLUGIN_FUNCTION (getMissingGpgKeyErrorText) (conf);
-		ELEKTRA_SET_INSTALLATION_ERROR (errorKey, errorDescription);
+		ELEKTRA_SET_VALIDATION_SEMANTIC_ERROR (errorKey, errorDescription);
 		elektraFree (errorDescription);
 		return -1;
 	}

--- a/src/plugins/crypto/openssl_operations.c
+++ b/src/plugins/crypto/openssl_operations.c
@@ -199,7 +199,7 @@ int elektraCryptoOpenSSLHandleCreate (elektraCryptoHandle ** handle, KeySet * co
 	{
 		keyDel (key);
 		keyDel (iv);
-		ELEKTRA_SET_VALIDATION_SYNTACTIC_ERROR (errorKey, "Failed to create handle! Invalid key length");
+		ELEKTRA_SET_VALIDATION_SEMANTIC_ERROR (errorKey, "Invalid key length was provided by the configuration");
 		return -1;
 	}
 
@@ -208,7 +208,7 @@ int elektraCryptoOpenSSLHandleCreate (elektraCryptoHandle ** handle, KeySet * co
 		keyDel (key);
 		keyDel (iv);
 
-		ELEKTRA_SET_VALIDATION_SYNTACTIC_ERROR (errorKey, "Failed to create handle! Invalid IV length");
+		ELEKTRA_SET_VALIDATION_SEMANTIC_ERROR (errorKey, "Invalid IV length was prodived by the configuration");
 		return -1;
 	}
 
@@ -240,8 +240,7 @@ int elektraCryptoOpenSSLHandleCreate (elektraCryptoHandle ** handle, KeySet * co
 
 	if (ERR_peek_error ())
 	{
-		// TODO: Correct??
-		ELEKTRA_SET_INTERNAL_ERRORF (errorKey, "Failed to create handle! libcrypto error code was: %lu", ERR_get_error ());
+		ELEKTRA_SET_INTERNAL_ERRORF (errorKey, "Failed to setup OpenSSL! libcrypto error code was: %lu", ERR_get_error ());
 		elektraFree (*handle);
 		*handle = NULL;
 		pthread_mutex_unlock (&mutex_ssl);
@@ -505,7 +504,6 @@ int elektraCryptoOpenSSLDecrypt (elektraCryptoHandle * handle, Key * k, Key * er
 	return 1;
 
 error:
-	// TODO: Correct??
 	ELEKTRA_SET_INTERNAL_ERRORF (errorKey, "Decryption error! libcrypto error code was: %lu", ERR_get_error ());
 	BIO_free_all (decrypted);
 	pthread_mutex_unlock (&mutex_ssl);
@@ -525,7 +523,6 @@ char * elektraCryptoOpenSSLCreateRandomString (Key * errorKey, const kdb_unsigne
 	pthread_mutex_lock (&mutex_ssl);
 	if (!RAND_bytes (buffer, length))
 	{
-		// TODO: Correct??
 		ELEKTRA_SET_INTERNAL_ERRORF (errorKey, "Failed to generate random string. libcrypto error code was: %lu", ERR_get_error ());
 		pthread_mutex_unlock (&mutex_ssl);
 		return NULL;

--- a/src/plugins/fcrypt/fcrypt.c
+++ b/src/plugins/fcrypt/fcrypt.c
@@ -127,10 +127,10 @@ static int shredTemporaryFile (int fd, Key * errorKey)
 
 	if (fstat (fd, &tmpStat))
 	{
-		ELEKTRA_SET_RESOURCE_ERRORF (
-			errorKey,
-			"Failed to overwrite the temporary data. Cannot retrieve file status. WARNING: Unencrypted data may leak. Errno: %s",
-			strerror (errno));
+		ELEKTRA_SET_RESOURCE_ERRORF (errorKey,
+					     "Failed to overwrite the temporary data. Cannot retrieve file status. WARNING: Unencrypted "
+					     "data may leak. Errno: %s",
+					     strerror (errno));
 		return -1;
 	}
 

--- a/src/plugins/fcrypt/fcrypt.c
+++ b/src/plugins/fcrypt/fcrypt.c
@@ -284,7 +284,7 @@ static int fcryptEncrypt (KeySet * pluginConfig, Key * parentKey)
 
 	if (recipientCount == 0 && signatureCount == 0)
 	{
-		ELEKTRA_SET_VALIDATION_SEMANTIC_ERROR (
+		ELEKTRA_SET_VALIDATION_SEMANTIC_ERRORF (
 			parentKey,
 			"Missing GPG recipient key (specified as %s) or GPG signature key (specified as %s) in plugin configuration",
 			ELEKTRA_RECIPIENT_KEY, ELEKTRA_SIGNATURE_KEY);

--- a/src/plugins/fcrypt/fcrypt.c
+++ b/src/plugins/fcrypt/fcrypt.c
@@ -129,7 +129,7 @@ static int shredTemporaryFile (int fd, Key * errorKey)
 	{
 		ELEKTRA_SET_RESOURCE_ERRORF (
 			errorKey,
-			"Failed to overwrite the temporary data. Cannot retrieve file status. Unencrypted data may leak. Errno: %s",
+			"Failed to overwrite the temporary data. Cannot retrieve file status. WARNING: Unencrypted data may leak. Errno: %s",
 			strerror (errno));
 		return -1;
 	}
@@ -149,7 +149,7 @@ static int shredTemporaryFile (int fd, Key * errorKey)
 	return 1;
 
 error:
-	ELEKTRA_SET_RESOURCE_ERRORF (errorKey, "Failed to overwrite the temporary data. Unencrypted data may leak. Reason: %s",
+	ELEKTRA_SET_RESOURCE_ERRORF (errorKey, "Failed to overwrite the temporary data. WARNING: Unencrypted data may leak! Reason: %s",
 				     strerror (errno));
 	return -1;
 }
@@ -251,7 +251,7 @@ static int fcryptGpgCallAndCleanup (Key * parentKey, KeySet * pluginConfig, char
 		{
 			ELEKTRA_ADD_RESOURCE_WARNINGF (
 				parentKey,
-				"Failed to unlink a temporary file. WARNING: unencrypted data may leak! Please try to delete "
+				"Failed to unlink a temporary file. WARNING: Unencrypted data may leak! Please try to delete "
 				"the file manually. Affected file: %s. Reason: %s",
 				tmpFile, strerror (errno));
 		}
@@ -284,7 +284,7 @@ static int fcryptEncrypt (KeySet * pluginConfig, Key * parentKey)
 
 	if (recipientCount == 0 && signatureCount == 0)
 	{
-		ELEKTRA_SET_INSTALLATION_ERRORF (
+		ELEKTRA_SET_VALIDATION_SEMANTIC_ERROR (
 			parentKey,
 			"Missing GPG recipient key (specified as %s) or GPG signature key (specified as %s) in plugin configuration",
 			ELEKTRA_RECIPIENT_KEY, ELEKTRA_SIGNATURE_KEY);
@@ -490,7 +490,7 @@ static int fcryptDecrypt (KeySet * pluginConfig, Key * parentKey, fcryptState * 
 		{
 			ELEKTRA_ADD_RESOURCE_WARNINGF (
 				parentKey,
-				"Failed to unlink a temporary file. WARNING: unencrypted data may leak! Please try to delete "
+				"Failed to unlink a temporary file. WARNING: Unencrypted data may leak! Please try to delete "
 				"the file manually. Affected file: %s, error description: %s",
 				tmpFile, strerror (errno));
 		}
@@ -607,7 +607,7 @@ int ELEKTRA_PLUGIN_FUNCTION (get) (Plugin * handle, KeySet * ks ELEKTRA_UNUSED, 
 			{
 				ELEKTRA_ADD_RESOURCE_WARNINGF (
 					parentKey,
-					"Failed to unlink a temporary file. WARNING: unencrypted data may leak! Please try "
+					"Failed to unlink a temporary file. WARNING: Unencrypted data may leak! Please try "
 					"to delete the file manually. Affected file: %s, error description: %s",
 					s->tmpFilePath, strerror (errno));
 			}
@@ -674,7 +674,7 @@ int ELEKTRA_PLUGIN_FUNCTION (checkconf) (Key * errorKey, KeySet * conf)
 	if (recipientCount == 0 && signatureCount == 0)
 	{
 		char * errorDescription = ELEKTRA_PLUGIN_FUNCTION (getMissingGpgKeyErrorText) (conf);
-		ELEKTRA_SET_INSTALLATION_ERROR (errorKey, errorDescription);
+		ELEKTRA_SET_VALIDATION_SEMANTIC_ERROR (errorKey, errorDescription);
 		elektraFree (errorDescription);
 		return -1;
 	}

--- a/src/plugins/gpgme/gpgme.c
+++ b/src/plugins/gpgme/gpgme.c
@@ -153,8 +153,8 @@ static gpgme_key_t * extractRecipientFromPluginConfig (KeySet * config, Key * er
 		err = gpgme_get_key (ctx, keyString (gpgRecipientRoot), &key, 0);
 		if (err)
 		{
-			ELEKTRA_SET_VALIDATION_SEMANTIC_ERROR (errorKey, "Failed to read the specified GPG key. Reason: %s",
-							       gpgme_strerror (err));
+			ELEKTRA_SET_VALIDATION_SEMANTIC_ERRORF (errorKey, "Failed to read the specified GPG key. Reason: %s",
+								gpgme_strerror (err));
 			elektraGpgmeKeylistFree (&list);
 			return NULL;
 		}
@@ -183,8 +183,8 @@ static gpgme_key_t * extractRecipientFromPluginConfig (KeySet * config, Key * er
 				err = gpgme_get_key (ctx, keyString (k), &key, 0);
 				if (err)
 				{
-					ELEKTRA_SET_VALIDATION_SEMANTIC_ERROR (errorKey, "Failed to read the specified GPG key. Reason: %s",
-									       gpgme_strerror (err));
+					ELEKTRA_SET_VALIDATION_SEMANTIC_ERRORF (
+						errorKey, "Failed to read the specified GPG key. Reason: %s", gpgme_strerror (err));
 					elektraGpgmeKeylistFree (&list);
 					return NULL;
 				}

--- a/src/plugins/gpgme/gpgme.c
+++ b/src/plugins/gpgme/gpgme.c
@@ -153,8 +153,7 @@ static gpgme_key_t * extractRecipientFromPluginConfig (KeySet * config, Key * er
 		err = gpgme_get_key (ctx, keyString (gpgRecipientRoot), &key, 0);
 		if (err)
 		{
-			// TODO: Correct??
-			ELEKTRA_SET_INTERNAL_ERRORF (errorKey, "Failed to receive the GPG key. Reason: %s", gpgme_strerror (err));
+			ELEKTRA_SET_VALIDATION_SEMANTIC_ERROR (errorKey, "Failed to read the specified GPG key. Reason: %s", gpgme_strerror (err));
 			elektraGpgmeKeylistFree (&list);
 			return NULL;
 		}
@@ -183,8 +182,7 @@ static gpgme_key_t * extractRecipientFromPluginConfig (KeySet * config, Key * er
 				err = gpgme_get_key (ctx, keyString (k), &key, 0);
 				if (err)
 				{
-					// TODO: Correct??
-					ELEKTRA_SET_INTERNAL_ERRORF (errorKey, "Failed to receive the GPG key. Reason: %s",
+					ELEKTRA_SET_VALIDATION_SEMANTIC_ERROR (errorKey, "Failed to read the specified GPG key. Reason: %s",
 								     gpgme_strerror (err));
 					elektraGpgmeKeylistFree (&list);
 					return NULL;
@@ -260,7 +258,7 @@ static int transferGpgmeDataToElektraKey (gpgme_data_t src, Key * dst, Key * err
 	readCount = gpgme_data_read (src, buffer, ciphertextLen);
 	if (readCount != ciphertextLen)
 	{
-		ELEKTRA_SET_INTERNAL_ERRORF (errorKey, "An error during occurred during the data transfer. Reason: %s", strerror (errno));
+		ELEKTRA_SET_INTERNAL_ERRORF (errorKey, "An error occurred during the en/decryption process. Reason: %s", strerror (errno));
 		returnValue = -1; // failure
 		goto cleanup;
 	}
@@ -351,7 +349,7 @@ static int gpgEncrypt (Plugin * handle, KeySet * data, Key * errorKey)
 	err = gpgme_new (&ctx);
 	if (err)
 	{
-		ELEKTRA_SET_INSTALLATION_ERRORF (errorKey, "Failed to create the gpgme context. Reason: %s", gpgme_strerror (err));
+		ELEKTRA_SET_INSTALLATION_ERRORF (errorKey, "Failed to initialize gpgme. Reason: %s", gpgme_strerror (err));
 		return -1; // at this point nothing has been initialized
 	}
 
@@ -486,7 +484,7 @@ static int gpgDecrypt (ELEKTRA_UNUSED Plugin * handle, KeySet * data, Key * erro
 	err = gpgme_new (&ctx);
 	if (err)
 	{
-		ELEKTRA_SET_INSTALLATION_ERRORF (errorKey, "Failed to the gpgme context. Reason: %s", gpgme_strerror (err));
+		ELEKTRA_SET_INSTALLATION_ERRORF (errorKey, "Failed to initialize gpgme. Reason: %s", gpgme_strerror (err));
 		return -1; // at this point nothing has been initialized
 	}
 
@@ -602,7 +600,7 @@ int elektraGpgmeCheckconf (Key * errorKey, KeySet * conf)
 	err = gpgme_new (&ctx);
 	if (err)
 	{
-		ELEKTRA_SET_INSTALLATION_ERRORF (errorKey, "Failed to create the gpgme context. Reason: %s", gpgme_strerror (err));
+		ELEKTRA_SET_INSTALLATION_ERRORF (errorKey, "Failed to initialize gpgme. Reason: %s", gpgme_strerror (err));
 		return -1; // at this point nothing has been initialized
 	}
 

--- a/src/plugins/gpgme/gpgme.c
+++ b/src/plugins/gpgme/gpgme.c
@@ -153,7 +153,8 @@ static gpgme_key_t * extractRecipientFromPluginConfig (KeySet * config, Key * er
 		err = gpgme_get_key (ctx, keyString (gpgRecipientRoot), &key, 0);
 		if (err)
 		{
-			ELEKTRA_SET_VALIDATION_SEMANTIC_ERROR (errorKey, "Failed to read the specified GPG key. Reason: %s", gpgme_strerror (err));
+			ELEKTRA_SET_VALIDATION_SEMANTIC_ERROR (errorKey, "Failed to read the specified GPG key. Reason: %s",
+							       gpgme_strerror (err));
 			elektraGpgmeKeylistFree (&list);
 			return NULL;
 		}
@@ -183,7 +184,7 @@ static gpgme_key_t * extractRecipientFromPluginConfig (KeySet * config, Key * er
 				if (err)
 				{
 					ELEKTRA_SET_VALIDATION_SEMANTIC_ERROR (errorKey, "Failed to read the specified GPG key. Reason: %s",
-								     gpgme_strerror (err));
+									       gpgme_strerror (err));
 					elektraGpgmeKeylistFree (&list);
 					return NULL;
 				}

